### ISSUE TITLE
KotlinClosure1 and similar can not specify null recivier

### ISSUE THREAD

**Issue by JLLeitschuh** [gradle/kotlin-dsl#593]
### Expected Behavior
Should be able to define a `KotlinClosure1` like this:
```kotlin
KotlinClosure1<Exception?, Unit> ({
    // Do something
}
```
### Current Behavior
Compiler error:
```
MessageBroker.gradle.kts:46:30: Type argument is not within its bounds: should be subtype of 'Any'
```

### Your Environment
```
------------------------------------------------------------
Gradle 4.3.1
------------------------------------------------------------

Build time:   2017-11-08 08:59:45 UTC
Revision:     e4f4804807ef7c2829da51877861ff06e07e006d

Groovy:       2.4.12
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_92 (Oracle Corporation 25.92-b14)
OS:           Mac OS X 10.12.6 x86_64

```

**Comment by JLLeitschuh** [comment]
This should fix the problem, changing `in T : Any` to `in T : Any?`.

```kotlin
class KotlinClosure1<in T : Any?, V : Any>(
    val function: T.() -> V?,
    owner: Any? = null,
    thisObject: Any? = null) : groovy.lang.Closure<V?>(owner, thisObject) {

    @Suppress("unused") // to be called dynamically by Groovy
    fun doCall(it: T): V? = it.function()
}
```

**Comment by JLLeitschuh** [comment]
While this problem is being solved, `V` should also be nullable as well.

**Comment by JLLeitschuh** [comment]
`V` seems to be handled as it is.

**Side Note**
I was kind of confused why this even compiled in Kotlin:
```kotlin
class KotlinClosure1<in T : Any, V : Any>(val function: T.() -> V?)
fun closure(function: String?.() -> String?) = KotlinClosure1(function)
```

Then I realized that it actually became:
```kotlin
fun closure(function: String?.() -> String?) = KotlinClosure1<String, String>(function)
```

PR will be forthcoming to fix this issue.

**Comment by bamboo** [comment]
Fixed in #594.
